okhttp/src/test/java/okhttp3/ConscryptTest.kt

```diff
@@ -15,34 +15,129 @@
 import okhttp3.TestUtil.assumeNetwork
 import okhttp3.internal.platform.ConscryptPlatform
 import okhttp3.internal.platform.Platform
 import okhttp3.testing.PlatformRule
+import okhttp3.tls.internal.TlsUtil
+import okio.ByteString.Companion.toByteString
 import org.assertj.core.api.Assertions.assertThat
 import org.conscrypt.Conscrypt
+import org.junit.jupiter.api.Assertions.assertEquals
 import org.junit.jupiter.api.Assertions.assertFalse
 import org.junit.jupiter.api.Assertions.assertTrue
 import org.junit.jupiter.api.BeforeEach
 import org.junit.jupiter.api.Disabled
 import org.junit.jupiter.api.Test
 import org.junit.jupiter.api.extension.RegisterExtension
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.ValueSource
+import javax.net.ssl.SSLSocket
 
 class ConscryptTest {
   @JvmField @RegisterExtension val platform = PlatformRule.conscrypt()
   @JvmField @RegisterExtension val clientTestRule = OkHttpClientTestRule()
 
-  private val client = clientTestRule.newClient()
+  private var client = clientTestRule.newClient()
 
-  @BeforeEach fun setUp() {
+  private val handshakeCertificates = TlsUtil.localhost()
+
+  private lateinit var server: MockWebServer
+
+  @BeforeEach @Throws(Exception::class) fun setUp(server: MockWebServer) {
     platform.assumeConscrypt()
+    this.server = server
+  }
+
+  private fun enableTls() {
+    client = client.newBuilder()
+      .sslSocketFactory(
+        handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager
+      )
+      .build()
+    server.useHttps(handshakeCertificates.sslSocketFactory(), false)
   }
 
   @Test
   fun testTrustManager() {
     assertThat(Conscrypt.isConscrypt(Platform.get().platformTrustManager())).isTrue()
   }
 
+  @ParameterizedTest(name = "{displayName}({arguments})")
+  @ValueSource(strings = ["TLSv1.2", "TLSv1.3"])
+  fun testSessionReuse(tlsVersion: String) {
+    val sessionIds = mutableListOf<String>()
+
+    enableTls()
+
+    val tlsVersion = TlsVersion.forJavaName(tlsVersion)
+    val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
+      .tlsVersions(tlsVersion)
+      .build()
+
+    var reuseSession = false
+
+    val sslContext = handshakeCertificates.sslContext()
+    val systemSslSocketFactory = sslContext.socketFactory
+    val sslSocketFactory = object : DelegatingSSLSocketFactory(systemSslSocketFactory) {
+      override fun configureSocket(sslSocket: SSLSocket): SSLSocket {
+        return sslSocket.apply {
+          if (reuseSession) {
+            this.enableSessionCreation = false
+          }
+        }
+      }
+    }
+
+    client = client.newBuilder().connectionSpecs(listOf(spec))
+      .eventListenerFactory(clientTestRule.wrap(object : EventListener() {
+        override fun connectionAcquired(call: Call, connection: Connection) {
+          val sslSocket = connection.socket() as SSLSocket
+
+          sessionIds.add(sslSocket.session.id.toByteString().hex())
```

Comment:
Omit the `.hex()` bit? Should still get a useful toString() if it fails.

Reply:
This is awkward without assertThat helpers for Okio.  Like isEmpty